### PR TITLE
[Core] Avoid uploading public key

### LINKED ISSUE #2025: [Spot/Auth] Public key should not be uploaded to the remote VM
Currently, we upload the ssh public key to the VM launched by SkyPilot for several clouds (Azure, Lambda, IBM, OCI, SCP): 
https://github.com/skypilot-org/skypilot/blob/8c497ed9b4f5bf871c613a54fc265febb2ae6354/sky/authentication.py#L306
https://github.com/skypilot-org/skypilot/blob/8c497ed9b4f5bf871c613a54fc265febb2ae6354/sky/authentication.py#L330
https://github.com/skypilot-org/skypilot/blob/8c497ed9b4f5bf871c613a54fc265febb2ae6354/sky/authentication.py#L388
https://github.com/skypilot-org/skypilot/blob/8c497ed9b4f5bf871c613a54fc265febb2ae6354/sky/authentication.py#L400

This will cause problem to the spot controller, as when the spot controller launch a new VM, a new ssh private key will be generated under `~/.ssh/sky-key`, while the public key is the one uploaded from the user's laptop `~/.ssh/sky-key.pub` which mismatch the generated one, causing the spot controller fail to log into the spot VM.

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #2025 

Uploading the `sky-key.pub` to the `~/.ssh/sky-key.pub` will cause problem for spot controller. This PR gets rid of the public_key upload, except for the Lambda, as it is hardcoded and hard to modify at the moment.

This also reduces the permission for the GCP account to launch a VM, as we no longer need to set the metadata for the key pairs.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch --cloud aws -c test-aws --num-nodes 2 -i 0 echo hi; sleep 90; sky status -r test-aws`
  - [x] `sky launch --cloud gcp -c test-gcp --num-nodes 2 -i 0 echo hi; sleep 90; sky status -r test-gcp` with a new pair of keys
  - [x] `sky launch --cloud gcp -c test-gcp -i 0; sky status -r` without the setCommonInstanceMetadata permission.
  - [x] `sky launch --cloud azure --num-nodes 2 -c test-azure -i 0 echo hi; sleep 120; sky status -r test-azure`
  - [x] `sky launch --cloud oci --num-nodes 2 --cpus 2 -c test-oci -i 0 echo hi; sleep 90; sky status -r test-oci`
  - [x] `sky launch --cloud lambda -c test-lambda -i 0 --num-nodes 2 --down echo hi; sleep 90; sky status -r test-lambda`
  - [x] `sky launch --cloud ibm --num-nodes 2 -c test-ibm -i 0 echo hi; sleep 90; sky status -r test-ibm`
  - [x] `sky spot launch --cpus 2 --cloud gcp echo hi` with Azure to be the spot controller #2025 
  - [ ] `sky launch --cloud scp -c test-scp -i 0 echo hi; sleep 90; sky status -r test-scp` (Failed due to test account not able to launch new instance)
- [x] All smoke tests: `pytest tests/test_smoke.py` (except #2207)

### CODE REVIEW

File: sky/skylet/providers/gcp/constants.py [hunk 1]
```diff
@@ -103,7 +103,6 @@
     "compute.subnetworks.list",
     "compute.subnetworks.useExternalIp",
     "compute.projects.get",
-    "compute.projects.setCommonInstanceMetadata",
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/constants.py:106]: Nice! Update permissions doc?

File: sky/skylet/providers/azure/config.py [hunk 2]
```diff
@@ -135,28 +135,7 @@ def _configure_resource_group(config):
...
-    for node_type in config["available_node_types"].values():
-        azure_arm_parameters = node_type["node_config"].setdefault(
-            "azure_arm_parameters", {}
-        )
-        azure_arm_parameters["adminUsername"] = ssh_user
-        azure_arm_parameters["publicKey"] = public_key
-
+    # SkyPilot: The original checks and configurations are no longer
+    # needed, since we have already set them up in the upper level
+    # SkyPilot codes.
```
Comment (concretevitamin) [inline sky/skylet/providers/azure/config.py:140]: Add: see sky/templates/azure-ray.yml.j2

File: sky/authentication.py [hunk 3]
```diff
@@ -267,71 +242,17 @@ def setup_gcp_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
             subprocess_utils.handle_returncode(proc.returncode, enable_ssh_cmd,
                                                'Failed to enable ssh port.',
                                                proc.stderr.decode('utf-8'))
-        return config
-
-    # OS Login is not enabled for the project. Add the ssh key directly to the
-    # metadata.
     # TODO(zhwu): Use cloud init to add ssh public key, to avoid the permission
```
Comment (concretevitamin) [inline sky/authentication.py:245]: Is this TODO still needed?

It mentions debian not having cloud-init.

File: sky/authentication.py [hunk 4]
```diff
@@ -422,14 +343,6 @@ def setup_oci_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
...
 
-    # Need to use ~ relative path because Ray uses the same
-    # path for finding the public key path on both local and head node.
-    config['auth']['ssh_public_key'] = PUBLIC_SSH_KEY_PATH
-
-    file_mounts = config['file_mounts']
-    file_mounts[PUBLIC_SSH_KEY_PATH] = PUBLIC_SSH_KEY_PATH
-    config['file_mounts'] = file_mounts
-
     for node_type in config['available_node_types']:
```
Comment (concretevitamin) [inline sky/authentication.py:346]: Add PR description to here?

File: sky/skylet/providers/lambda_cloud/lambda_utils.py [hunk 5]
```diff
@@ -189,7 +192,7 @@ def get_unique_ssh_key_name(self, prefix: str,
         suffix_digits = [0]
         for key_info in candidate_keys:
             name = key_info.get('name', '')
-            if key_info.get('public_key', '') == pub_key:
+            if key_info.get('public_key', '').strip() == pub_key.strip():
```
Comment (concretevitamin) [inline sky/skylet/providers/lambda_cloud/lambda_utils.py:195]: What bug does this fix?

File: sky/skylet/providers/lambda_cloud/lambda_utils.py [hunk 6]
```diff
@@ -4,8 +4,11 @@
 import requests
 from typing import Any, Dict, List, Optional, Tuple
 
+from sky.utils import common_utils
+
 CREDENTIALS_PATH = '~/.lambda_cloud/lambda_keys'
 API_ENDPOINT = 'https://cloud.lambdalabs.com/api/v1'
+SSH_KEY_PREFIX = f'sky-key-{common_utils.get_user_hash()}'
```
Comment (concretevitamin) [inline sky/skylet/providers/lambda_cloud/lambda_utils.py:11]: Unused?

File: sky/skylet/providers/scp/config.py [hunk 7]
```diff
@@ -50,7 +50,7 @@ def bootstrap_instance_config(self, node_config):
         instance_config['serverType'] = node_config['InstanceType']
         instance_config['contractId'] = "None"
         instance_config['initialScript'] = self._get_vm_init_script(
-            node_config['auth']['ssh_public_key'])
+            node_config['AuthorizedKey'])
```
Comment (concretevitamin) [inline sky/skylet/providers/scp/config.py:53]: Did we pick AuthorizedKey on our own, or is it a convention?

File: sky/templates/aws-ray.yml.j2 [hunk 8]
```diff
@@ -74,7 +74,8 @@ available_node_types:
             shell: /bin/bash
             sudo: ALL=(ALL) NOPASSWD:ALL
             ssh_authorized_keys:
-              - skypilot:ssh_public_key_content
+              - |
```
Comment (concretevitamin) [inline sky/templates/aws-ray.yml.j2:77]: What does this change do?

File: sky/templates/gcp-ray.yml.j2 [hunk 9]
```diff
@@ -59,8 +59,13 @@ available_node_types:
       guestAccelerators:
         - acceleratorType: projects/{{gcp_project_id}}/zones/{{zones}}/acceleratorTypes/{{gpu}}
           acceleratorCount: {{gpu_count}}
+  {%- endif %}
       metadata:
         items:
+          - key: ssh-keys
+            value: |
+              skypilot:ssh_user:skypilot:ssh_public_key_content
```
Comment (concretevitamin) [inline sky/templates/gcp-ray.yml.j2:67]: After replacement does this become `me:<my pub key content>`? Is this the expected syntax? Worth a comment.

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll, left some questions. We should probably do back compat + smoke.

**Comment by Michaelvll** [comment]
Thanks for the review @concretevitamin! It passed the smoke tests. For backward compatibility, the changes in this PR will not affect the existing clusters, as the config restoring in `write_cluster_config` will not apply any of the new changes to the existing cluster.

File: sky/templates/aws-ray.yml.j2 (hunk 8 above)
Comment (concretevitamin) [reply sky/templates/aws-ray.yml.j2:77]: Worth copying this into a comment!

File: sky/templates/lambda-ray.yml.j2 [hunk 10]
```diff
@@ -24,13 +24,17 @@ available_node_types:
   ray_head_default:
     resources: {}
     node_config:
+      AuthorizedKey: |
```
Comment (concretevitamin) [inline sky/templates/lambda-ray.yml.j2:27]: Is this interpreted/used anywhere?

File: sky/skylet/providers/scp/config.py (hunk 7 above)
Comment (concretevitamin) [reply sky/skylet/providers/scp/config.py:53]: I guess my question is it seems a bit weird that the different *.j2 use different field names for this.

One option is to conform to AWS's j2, i.e., cloud-init's "ssh_authorized_keys" field?

**Review by concretevitamin — COMMENTED** [review]
Thanks, some minor questions.

File: sky/skylet/providers/scp/config.py (hunk 7 above)
Comment (Michaelvll) [reply sky/skylet/providers/scp/config.py:53]: I think the problem is that we want to keep the naming style the same as the other keys in the yaml file. For example, in `oci-ray.yml.j2`, all the keys are camel-case. Having a snake-case `ssh_authorized_keys` among them might be a bit weird?

File: sky/templates/aws-ray.yml.j2 (hunk 8 above)
Comment (Michaelvll) [reply sky/templates/aws-ray.yml.j2:80]: Done. Thanks!

File: sky/templates/lambda-ray.yml.j2 (hunk 10 above)
Comment (Michaelvll) [reply sky/templates/lambda-ray.yml.j2]: Good catch! That is a remnant. Removed. Thanks!

File: sky/authentication.py [hunk 11]
```diff
@@ -1,4 +1,21 @@
-"""Module to enable a single SkyPilot key for all VMs in each cloud."""
+"""Module to enable a single SkyPilot key for all VMs in each cloud.
+
+The `setup_<cloud>_authentication` functions will be called after the ray
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/authentication.py]: The functions will be called no matter the cluster exists or not, although it does not affect the existing cluster. We can fix it in a future PR. : )

File: sky/authentication.py [hunk 12]
```diff
@@ -1,20 +1,25 @@
...
-yaml template file `<cloud>-ray.yml.j2` is filled in with resource specific
-information. The functions will take the filled in ray yaml config as input,
+The `setup_<cloud>_authentication` functions will be called on every cluster
+provisioning request.
+
+Specifically, after the ray yaml template file `<cloud>-ray.yml.j2` is filled in
+with resource specific information, these functions are called with the filled
+in ray yaml config as input, yaml template file `<cloud>-ray.yml.j2` is filled
+in with resource specific information. The functions will take the filled in ray
+yaml config as input,
```
Comment (concretevitamin) [inline sky/authentication.py:10]: some duplication

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks @Michaelvll!

File: sky/authentication.py (hunk 12 above)
Comment (Michaelvll) [reply sky/authentication.py]: Oops, the applying of the changes was buggy. Fixed. Thanks!

**Pull request merged**